pool.go

```diff
@@ -0,0 +1,192 @@
+package server
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/bblfsh/sdk/protocol"
+	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
+)
+
+const (
+	DefaultPoolMin               = 1
+	DefaultPoolMax               = 5
+	DefaultPoolTimeBeforeNew     = time.Second * 3
+	DefaultPoolTimeBetweenSpawns = time.Second * 10
+	DefaultPoolTimeBeforeClean   = time.Minute * 1
+)
+
+// DriverPool controls a pool of drivers and balances requests among them,
+// ensuring each driver does not get concurrent requests.
+type DriverPool struct {
```

Comment:
I would add an interface for the driver pool, so the main public functionality can be understood.  Then, I would use this as an implementation (with the internal scaler and everything explained in a completely different context than the driver main functionality).  That way people will be able to tell what attributes are there to help fulfilling the interface (example `New`) and which attributes are there to help with the implementation (`TimeBeforeClean`).